torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -2899,10 +2901,20 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
                     ), "Currently the way for _is_sharded to be False is \
                         world_size == 1 or sharding_stratagy is set to be NO_SHARD"
                     if self.sharding_strategy == ShardingStrategy.NO_SHARD:
-                        dist.all_reduce(param.grad, group=self.process_group)
-                        if self.gradient_postdivide_factor > 1:
-                            # Average grad by world_size for consistency with PyTorch DDP.
-                            param.grad.div_(self.gradient_postdivide_factor)
+                        # if a communication hook was registered,
+                        # then a model has attribute ``communication_hook``
+                        if hasattr(self, "communication_hook"):
+                            self.communication_hook(self.communication_hook_state, param.grad)
+                        else:
```

Comment:
This branch can be put into a default hook, so that two branch can merge into one?

Reply:
Thank you for suggestion! Refactored code accordingly and added some tests